bindings/python/pymongoarrow/builder.pyx

```diff
@@ -0,0 +1,94 @@
+# Copyright 2021-present MongoDB, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Cython compiler directives
+# distutils: language=c++
+# cython: language_level=3
+from pyarrow.lib cimport *
+
+import numpy as np
+
+
+cdef class Int32Builder(_Weakrefable):
+    cdef:
+        shared_ptr[CInt32Builder] builder
+
+    def __cinit__(self, MemoryPool memory_pool=None):
+        cdef CMemoryPool* pool = maybe_unbox_memory_pool(memory_pool)
+        self.builder.reset(new CInt32Builder(pool))
+
+    def append(self, value):
+        if value is None or value is np.nan:
+            self.builder.get().AppendNull()
+        elif isinstance(value, int):
+            self.builder.get().Append(value)
+        else:
+            raise TypeError('Int32Builder only accepts integer objects')
```

Comment:
Instead of calling `isinstance(value, int)`, can we rely on CInt32Builder to raise an error when the value is not supported? If so it would have better performance.

Reply:
We aren't going to be using this method to actually construct the arrays from BSON streams. That code will be pure-C/C++. These wrapper classes will help us create the necessary builders etc from Python code instead of needing to do this via C++ code which is rather complex due to scoping rules enforced by Cython. 

See the `unwrap` method in all these classes - this will eventually be used by some Cython code to access the underlying builders directly to construct arrays. The need for an unwrap method is also why the underlying builder classes are `cdef`ined as `shared_ptr` instances instead of `unique_ptr` instances (so that they can be unwrapped and a new reference to them can be passed to pure-C code somewhere).